plugins/inputs/http_listener_v2/http_listener_v2.go

```diff
@@ -0,0 +1,304 @@
+	h.mu.Lock()
+	defer h.mu.Unlock()
+
+	if h.MaxBodySize == 0 {
+		h.MaxBodySize = defaultMaxBodySize
+	}
+
+	if h.ReadTimeout.Duration < time.Second {
+		h.ReadTimeout.Duration = time.Second * 10
+	}
+	if h.WriteTimeout.Duration < time.Second {
+		h.WriteTimeout.Duration = time.Second * 10
+	}
+
+	h.acc = acc
+
+	tlsConf, err := h.ServerConfig.TLSConfig()
+	if err != nil {
+		return err
+	}
+
+	server := &http.Server{
+		Addr:         h.ServiceAddress,
+		Handler:      h,
+		ReadTimeout:  h.ReadTimeout.Duration,
+		WriteTimeout: h.WriteTimeout.Duration,
+		TLSConfig:    tlsConf,
+	}
+
+	var listener net.Listener
+	if tlsConf != nil {
+		listener, err = tls.Listen("tcp", h.ServiceAddress, tlsConf)
+	} else {
+		listener, err = net.Listen("tcp", h.ServiceAddress)
+	}
+	if err != nil {
+		return err
+	}
+	h.listener = listener
+	h.Port = listener.Addr().(*net.TCPAddr).Port
+
+	h.wg.Add(1)
+	go func() {
+		defer h.wg.Done()
+		server.Serve(h.listener)
+	}()
+
+	log.Printf("I! Started HTTP listener V2 service on %s\n", h.ServiceAddress)
+
+	return nil
+}
+
+// Stop cleans up all resources
+func (h *HTTPListenerV2) Stop() {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+
+	h.listener.Close()
+	h.wg.Wait()
+
+	log.Println("I! Stopped HTTP listener V2 service on ", h.ServiceAddress)
+}
+
+func (h *HTTPListenerV2) ServeHTTP(res http.ResponseWriter, req *http.Request) {
+	if req.URL.Path == h.Path {
+		h.AuthenticateIfSet(h.serveWrite, res, req)
+	} else {
+		h.AuthenticateIfSet(http.NotFound, res, req)
+	}
+}
+
+func (h *HTTPListenerV2) serveWrite(res http.ResponseWriter, req *http.Request) {
+	// Check that the content length is not too large for us to handle.
+	if req.ContentLength > h.MaxBodySize {
+		tooLarge(res)
+		return
+	}
+
+	// Check if the requested HTTP method was specified in config.
+	isAcceptedMethod := false
+	for _, method := range h.Methods {
+		if req.Method == method {
+			isAcceptedMethod = true
+			break
+		}
+	}
+	if !isAcceptedMethod {
+		methodNotAllowed(res)
+		return
+	}
+
+	// Handle gzip request bodies
+	body := req.Body
+	if req.Header.Get("Content-Encoding") == "gzip" {
+		var err error
+		body, err = gzip.NewReader(req.Body)
+		defer body.Close()
+		if err != nil {
+			log.Println("D! " + err.Error())
+			badRequest(res, err.Error())
+			return
+		}
+	}
+
+	// Add +1 for EOF
+	buf := make([]byte, h.MaxBodySize+1)
+	n, err := io.ReadFull(body, buf)
+	if err == io.ErrUnexpectedEOF || err == io.EOF {
+		// finished reading the request body
+		if err := h.parse(buf[:n]); err != nil {
+			log.Println("D! " + err.Error())
+			badRequest(res, err.Error())
+		} else {
+			res.WriteHeader(http.StatusNoContent)
+		}
+	} else {
+		log.Println("D! " + err.Error())
+		// problem reading the request body
+		internalServerError(res)
+	}
+}
+
+func (h *HTTPListenerV2) parse(b []byte) error {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+
+	metrics, err := h.Parse(b)
+	if err != nil {
+		return fmt.Errorf("unable to parse: %s", err.Error())
```

Comment:
Don't add the extra error context here, this is done in the other plugin only to mirror InfluxDB's error messages.